wgpu-hal/src/lib.rs

```diff
@@ -190,50 +191,196 @@ impl InstanceError {
+    fn as_any_mut(&mut self) -> &mut dyn Any;
+}
+impl<T: WasmNotSendSync + 'static> Resource for T {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+    fn as_any_mut(&mut self) -> &mut dyn Any {
+        self
+    }
+}
 
-    type Buffer: fmt::Debug + WasmNotSendSync + 'static;
-    type Texture: fmt::Debug + WasmNotSendSync + 'static;
-    type SurfaceTexture: fmt::Debug + WasmNotSendSync + Borrow<Self::Texture>;
-    type TextureView: fmt::Debug + WasmNotSendSync;
-    type Sampler: fmt::Debug + WasmNotSendSync;
-    type QuerySet: fmt::Debug + WasmNotSendSync;
-    type Fence: fmt::Debug + WasmNotSendSync;
+resource!(CommandBuffer: fmt::Debug);
 
-    type BindGroupLayout: fmt::Debug + WasmNotSendSync;
-    type BindGroup: fmt::Debug + WasmNotSendSync;
-    type PipelineLayout: fmt::Debug + WasmNotSendSync;
-    type ShaderModule: fmt::Debug + WasmNotSendSync;
-    type RenderPipeline: fmt::Debug + WasmNotSendSync;
-    type ComputePipeline: fmt::Debug + WasmNotSendSync;
+resource!(Buffer: fmt::Debug + 'static);
+resource!(Texture: fmt::Debug + 'static);
+resource!(TextureView: fmt::Debug);
+resource!(Sampler: fmt::Debug);
+resource!(QuerySet: fmt::Debug);
+resource!(Fence: fmt::Debug);
 
-    type AccelerationStructure: fmt::Debug + WasmNotSendSync + 'static;
-}
+resource!(BindGroupLayout: fmt::Debug);
+resource!(BindGroup: fmt::Debug);
+resource!(PipelineLayout: fmt::Debug);
+resource!(ShaderModule: fmt::Debug);
+resource!(RenderPipeline: fmt::Debug);
+resource!(ComputePipeline: fmt::Debug);
+
+resource!(AccelerationStructure: fmt::Debug + 'static);
 
-pub trait Instance: Sized + WasmNotSendSync {
-    type A: Api;
+pub trait SurfaceTexture<T: Texture + ?Sized>: Resource + fmt::Debug + Borrow<T> {}
+impl<T: Resource + fmt::Debug + Borrow<Tex>, Tex: Texture> SurfaceTexture<Tex> for T {}
 
-    unsafe fn init(desc: &InstanceDescriptor) -> Result<Self, InstanceError>;
+pub trait Api: Clone + fmt::Debug + Sized {
+    type Instance: Instance<Self> + ?Sized;
+    type Surface: Surface<Self> + ?Sized;
+    type Adapter: Adapter<Self> + ?Sized;
+    type Device: Device<Self> + ?Sized;
+
+    type Queue: Queue<Self> + ?Sized;
+    type CommandEncoder: CommandEncoder<Self> + ?Sized;
+    type CommandBuffer: CommandBuffer + ?Sized;
+
+    type Buffer: Buffer + ?Sized;
+    type Texture: Texture + ?Sized;
+    type SurfaceTexture: SurfaceTexture<Self::Texture> + ?Sized;
+    type TextureView: TextureView + ?Sized;
+    type Sampler: Sampler + ?Sized;
+    type QuerySet: QuerySet + ?Sized;
+    type Fence: Fence + ?Sized;
+
+    type BindGroupLayout: BindGroupLayout + ?Sized;
+    type BindGroup: BindGroup + ?Sized;
+    type PipelineLayout: PipelineLayout + ?Sized;
+    type ShaderModule: ShaderModule + ?Sized;
+    type RenderPipeline: RenderPipeline + ?Sized;
+    type ComputePipeline: ComputePipeline + ?Sized;
+
+    type AccelerationStructure: AccelerationStructure + ?Sized;
+}
+
+pub trait ConcreteApi:
```

Comment:
I wanted to keep `Api` as is because it prevents a breaking change. I considered `SizedApi`, but I thought `ConcreteApi` sounded nicer.

Do note that the associated types on `ConcreteApi` are pretty much just an implementation detail. Due to the `=` bound, the regular `Api` associated types can be used in the same places.

Reply:
@Wumpf what's your position on this? Are you alright with the current naming, or do you have another proposal?

Reply:
let's keep the current naming